### ISSUE TITLE
Make it easier to override the HTTPSender in azblob

### ISSUE THREAD

**Issue by ncw** [Azure/azure-storage-blob-go#89]
### Which version of the SDK was used?

cf01652132cc

### Which platform are you using? (ex: Windows, Linux, Debian)

Linux

### What problem was encountered?

It is very hard to override the HTTPSender when using azureblob.

It means copying this code and modifying it to create the pipeline.  Copying that code means that I won't get further updates to the pipeline that the Azure team make.  Also `anonymousCredentialPolicyFactory` is private so I can't do that bit.

https://github.com/Azure/azure-storage-blob-go/blob/cf01652132ccc92af2b9da9d2de542b5b7eafc1d/azblob/zc_pipeline.go#L22-L42

I'd like to propose this patch as a potential solution, but I thought I'd open an issue to discuss before I send a PR.

```diff
diff --git a/vendor/github.com/Azure/azure-storage-blob-go/azblob/zc_pipeline.go b/vendor/github.com/Azure/azure-storage-blob-go/azblob/zc_pipeline.go
index f34cd0a7b..91db40a99 100644
--- a/vendor/github.com/Azure/azure-storage-blob-go/azblob/zc_pipeline.go
+++ b/vendor/github.com/Azure/azure-storage-blob-go/azblob/zc_pipeline.go
@@ -17,6 +17,10 @@ type PipelineOptions struct {
 
 	// Telemetry configures the built-in telemetry policy behavior.
 	Telemetry TelemetryOptions
+
+	// HTTPSender allow the sender to be overidden. If it is nil, then
+	// the pipeline's default client is used to send the HTTP requests.
+	HTTPSender Factory
 }
 
 // NewPipeline creates a Pipeline using the specified credentials and options.
@@ -38,5 +42,5 @@ func NewPipeline(c Credential, o PipelineOptions) pipeline.Pipeline {
 		pipeline.MethodFactoryMarker(), // indicates at what stage in the pipeline the method factory is invoked
 		NewRequestLogPolicyFactory(o.RequestLog))
 
-	return pipeline.NewPipeline(f, pipeline.Options{HTTPSender: nil, Log: o.Log})
+	return pipeline.NewPipeline(f, pipeline.Options{HTTPSender: o.HTTPSender, Log: o.Log})
 }
```

**Comment by rickle-msft** [comment]
Hi, @ncw. Thank you for proposing this change. We will discuss this as a team and let you know if we will be able to take this enhancement.

**Comment by JeffreyRichter** [comment]
anonymousCredentialFactory should remain private; instead, call the public NewAnonymousCredential() function. Then, I'd add a pipeline.Options field to the PipelineOptions struct and if it's Log field is null, set it to o.Log. What I describe to not totally ideal (passing a log callback twice in the same structure) but it is a non-breaking change.

**Comment by zezha-msft** [comment]
@JeffreyRichter having a `pipeline.Options` inside the `PipelineOptions` struct sounds really confusing for the users. In addition, there's two log callbacks like you said. What do you think of @ncw's suggestion to add `HTTPSender` to `PipelineOptions`?

**Comment by JeffreyRichter** [comment]
OK, I just re-looked over the code and yes, I think you can just add HTTPSender (a pipeline.Factory) field to azblob's PipelineOptions struct. There are already 2 log-related fields in this same struct but they do different things (which is OK). Then I suggest making the same change to the Files & Queues SDKs.

**Comment by ncw** [comment]
Would you like me to send a PR for the above?

**Comment by zezha-msft** [comment]
@ncw that'd be awesome! Please let me know if you need any help or have any question.

**Comment by sumit-kalra** [comment]
@ncw I need the exact same functionality so I am curious - are you still working on this PR?

**Comment by ncw** [comment]
@sumit-kalra wrote:
>  I need the exact same functionality so I am curious - are you still working on this PR?

No, it dropped off my radar!  Please feel free to tidy the code up and send a PR

**Comment by zezha-msft** [comment]
Hi @sumit-kalra @ncw it's been shipped in `v0.6.0`. Thank you!
